ql/src/java/org/apache/hadoop/hive/ql/parse/type/RexNodeExprFactory.java

```diff
@@ -0,0 +1,1021 @@
+        makeHiveUnicodeString(Interpretation.STRING, value));
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  protected RexLiteral createDateConstantExpr(String value) {
+    Date d = Date.valueOf(value);
+    return rexBuilder.makeDateLiteral(
+        DateString.fromDaysSinceEpoch(d.toEpochDay()));
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  protected RexLiteral createTimestampConstantExpr(String value) {
+    Timestamp t = Timestamp.valueOf(value);
+    return (RexLiteral) rexBuilder.makeLiteral(
+        TimestampString.fromMillisSinceEpoch(t.toEpochMilli()).withNanos(t.getNanos()),
+        rexBuilder.getTypeFactory().createSqlType(
+            SqlTypeName.TIMESTAMP,
+            rexBuilder.getTypeFactory().getTypeSystem().getDefaultPrecision(SqlTypeName.TIMESTAMP)),
+        false);
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  protected RexLiteral createTimestampLocalTimeZoneConstantExpr(String value, ZoneId zoneId) {
+    TimestampTZ t = TimestampTZUtil.parse(value);
+
+    final TimestampString tsLocalTZString;
+    if (value == null) {
+      tsLocalTZString = null;
+    } else {
+      Instant i = t.getZonedDateTime().toInstant();
+      tsLocalTZString = TimestampString
+          .fromMillisSinceEpoch(i.toEpochMilli())
+          .withNanos(i.getNano());
+    }
+    return rexBuilder.makeTimestampWithLocalTimeZoneLiteral(
+        tsLocalTZString,
+        rexBuilder.getTypeFactory().getTypeSystem().getDefaultPrecision(SqlTypeName.TIMESTAMP_WITH_LOCAL_TIME_ZONE));
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  protected RexLiteral createIntervalYearMonthConstantExpr(String value) {
+    BigDecimal totalMonths = BigDecimal.valueOf(HiveIntervalYearMonth.valueOf(value).getTotalMonths());
+    return rexBuilder.makeIntervalLiteral(totalMonths,
+        new SqlIntervalQualifier(TimeUnit.YEAR, TimeUnit.MONTH, new SqlParserPos(1, 1)));
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  protected RexLiteral createIntervalDayTimeConstantExpr(String value) {
+    HiveIntervalDayTime v = HiveIntervalDayTime.valueOf(value);
+    BigDecimal secsValueBd = BigDecimal
+        .valueOf(v.getTotalSeconds() * 1000);
+    BigDecimal nanosValueBd = BigDecimal.valueOf((v).getNanos(), 6);
+    return rexBuilder.makeIntervalLiteral(secsValueBd.add(nanosValueBd),
+        new SqlIntervalQualifier(TimeUnit.MILLISECOND, null, new
+            SqlParserPos(1, 1)));
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  protected RexLiteral createIntervalYearConstantExpr(String value) {
+    HiveIntervalYearMonth v = new HiveIntervalYearMonth(Integer.parseInt(value), 0);
+    BigDecimal totalMonths = BigDecimal.valueOf(v.getTotalMonths());
+    return rexBuilder.makeIntervalLiteral(totalMonths,
+        new SqlIntervalQualifier(TimeUnit.YEAR, TimeUnit.MONTH, new SqlParserPos(1, 1)));
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  protected RexLiteral createIntervalMonthConstantExpr(String value) {
```

Comment:
I think you are right that this can be refined further. To be honest, my main goal here was just to mimic existing interface (which was based on already existing calls) and separate concepts without refactoring more than necessary... and still look at this jumbo patch. However, I think you are right. Maybe the factory should be split into smaller classes... Or maybe we should create a single method and pass information as a parameter. I have created https://issues.apache.org/jira/browse/HIVE-23225.